SecurityFilterMiddleware executes side-effects outside IO

Consider this code (http4s-pac4j 1.0.1):
```scala
import cats.effect.IO
import org.http4s._
import org.pac4j.http4s._
import org.http4s.dsl.io._
import org.pac4j.core.config.Config
import org.pac4j.oidc.config.OidcConfiguration
import org.pac4j.oidc.client.OidcClient

object Foo extends App {
  val oidc = new OidcConfiguration
  oidc.setDiscoveryURI("http://localhost:1234")
  oidc.setClientId("foo")
  oidc.setSecret("foo")

  val client = new OidcClient(oidc)
  client.setCallbackUrl("http://localhost:1234")

  val config = new Config(client)
  config.setHttpActionAdapter(DefaultHttpActionAdapter)
  config.setSessionStore(Http4sCookieSessionStore)
  val filter = SecurityFilterMiddleware.securityFilter(config)

  val okRoutes = HttpRoutes.of[IO] { case _ => Ok() }

  filter(okRoutes).run(Request(GET, Uri()))
}
```
By all rights, since we don't actually `.unsafeRunSync` the resulting `IO`, this program should do nothing. But if we run it, we get an exception:
```
org.pac4j.core.exception.TechnicalException: java.net.ConnectException: Connection refused (Connection refused)
	at org.pac4j.oidc.config.OidcConfiguration.internalInit(OidcConfiguration.java:147)
	at org.pac4j.core.util.InitializableObject.init(InitializableObject.java:20)
	at org.pac4j.oidc.client.OidcClient.clientInit(OidcClient.java:45)
	at org.pac4j.core.client.IndirectClient.internalInit(IndirectClient.java:58)
	at org.pac4j.core.util.InitializableObject.init(InitializableObject.java:20)
	at org.pac4j.core.client.IndirectClient.getRedirectAction(IndirectClient.java:93)
	at org.pac4j.core.client.IndirectClient.redirect(IndirectClient.java:79)
	at org.pac4j.core.engine.DefaultSecurityLogic.redirectToIdentityProvider(DefaultSecurityLogic.java:217)
	at org.pac4j.core.engine.DefaultSecurityLogic.perform(DefaultSecurityLogic.java:149)
	at org.pac4j.http4s.SecurityFilterMiddleware$.$anonfun$securityFilter$1(SecurityFilterMiddleware.scala:69)
	at org.http4s.server.package$Middleware$.$anonfun$apply$2(package.scala:64)
	at Foo$.delayedEndpoint$Foo$1(Foo.scala:25)
	at Foo$delayedInit$body.apply(Foo.scala:9)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
	at scala.App.$anonfun$main$1$adapted(App.scala:80)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.App.main(App.scala:80)
	at scala.App.main$(App.scala:78)
	at Foo$.main(Foo.scala:9)
	at Foo.main(Foo.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
	at java.base/java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.base/java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:399)
	at java.base/java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:242)
	at java.base/java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:224)
	at java.base/java.net.Socket.connect(Socket.java:609)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:177)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:474)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:569)
	at java.base/sun.net.www.http.HttpClient.<init>(HttpClient.java:242)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:341)
	at java.base/sun.net.www.http.HttpClient.New(HttpClient.java:362)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1253)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1187)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$6.run(HttpURLConnection.java:1071)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$6.run(HttpURLConnection.java:1069)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.AccessController.doPrivilegedWithCombiner(AccessController.java:795)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1068)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:1015)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1592)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$9.run(HttpURLConnection.java:1512)
	at java.base/sun.net.www.protocol.http.HttpURLConnection$9.run(HttpURLConnection.java:1510)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.AccessController.doPrivilegedWithCombiner(AccessController.java:795)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1509)
	at com.nimbusds.jose.util.DefaultResourceRetriever.getInputStream(DefaultResourceRetriever.java:249)
	at com.nimbusds.jose.util.DefaultResourceRetriever.retrieveResource(DefaultResourceRetriever.java:201)
	at org.pac4j.oidc.config.OidcConfiguration.internalInit(OidcConfiguration.java:144)
	at org.pac4j.core.util.InitializableObject.init(InitializableObject.java:20)
	at org.pac4j.oidc.client.OidcClient.clientInit(OidcClient.java:45)
	at org.pac4j.core.client.IndirectClient.internalInit(IndirectClient.java:58)
	at org.pac4j.core.util.InitializableObject.init(InitializableObject.java:20)
	at org.pac4j.core.client.IndirectClient.getRedirectAction(IndirectClient.java:93)
	at org.pac4j.core.client.IndirectClient.redirect(IndirectClient.java:79)
	at org.pac4j.core.engine.DefaultSecurityLogic.redirectToIdentityProvider(DefaultSecurityLogic.java:217)
	at org.pac4j.core.engine.DefaultSecurityLogic.perform(DefaultSecurityLogic.java:149)
	at org.pac4j.http4s.SecurityFilterMiddleware$.$anonfun$securityFilter$1(SecurityFilterMiddleware.scala:69)
	at org.http4s.server.package$Middleware$.$anonfun$apply$2(package.scala:64)
	at Foo$.delayedEndpoint$Foo$1(Foo.scala:25)
	at Foo$delayedInit$body.apply(Foo.scala:9)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at scala.Function0.apply$mcV$sp$(Function0.scala:39)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:17)
	at scala.App.$anonfun$main$1$adapted(App.scala:80)
	at scala.collection.immutable.List.foreach(List.scala:431)
	at scala.App.main(App.scala:80)
	at scala.App.main$(App.scala:78)
	at Foo$.main(Foo.scala:9)
	at Foo.main(Foo.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```
The reason looks to be `securityLogic.perform` [here](https://github.com/pac4j/http4s-pac4j/blob/1e026a7f3595e9a944754474bc119fba77fe2af9/src/main/scala/org/pac4j/http4s/SecurityFilterMiddleware.scala#L60), which is run outside `IO`.

--- a/src/main/scala/org/pac4j/http4s/CallbackService.scala
+++ b/src/main/scala/org/pac4j/http4s/CallbackService.scala
@@ -23,12 +23,12 @@ class CallbackService(config: Config,
                       multiProfile: Boolean = false,
                       renewSession: Boolean = true,
                       defaultClient: Option[String] = None)
-                      (implicit cs: ContextShift[IO], sync: Sync[IO]){
+                      (implicit cs: ContextShift[IO]){
 
   def login(request: Request[IO]): IO[Response[IO]] = {
     val callbackLogic = new DefaultCallbackLogic[IO[Response[IO]], Http4sWebContext]()
     val webContext = Http4sWebContext(request, config)
-    blocker.delay(callbackLogic.perform(webContext,
+    blocker.delay[IO, IO[Response[IO]]](callbackLogic.perform(webContext,
       config,
       config.getHttpActionAdapter.asInstanceOf[HttpActionAdapter[IO[Response[IO]], Http4sWebContext]],
       this.defaultUrl.orNull,

--- a/src/main/scala/org/pac4j/http4s/CallbackService.scala
+++ b/src/main/scala/org/pac4j/http4s/CallbackService.scala
@@ -1,6 +1,7 @@
 package org.pac4j.http4s
 
-import cats.effect.IO
+import cats.implicits._
+import cats.effect._
 import org.http4s.{Request, Response}
 import org.pac4j.core.config.Config
 import org.pac4j.core.engine.DefaultCallbackLogic
@@ -16,22 +17,24 @@ import org.pac4j.core.http.adapter.HttpActionAdapter
   * @author Iain Cardnell
   */
 class CallbackService(config: Config,
+                      blocker: Blocker,
                       defaultUrl: Option[String] = None,
                       saveInSession: Boolean = true,
                       multiProfile: Boolean = false,
                       renewSession: Boolean = true,
-                      defaultClient: Option[String] = None) {
+                      defaultClient: Option[String] = None)
+                      (implicit cs: ContextShift[IO], sync: Sync[IO]){
 
   def login(request: Request[IO]): IO[Response[IO]] = {
     val callbackLogic = new DefaultCallbackLogic[IO[Response[IO]], Http4sWebContext]()
     val webContext = Http4sWebContext(request, config)
-    callbackLogic.perform(webContext,
+    blocker.delay(callbackLogic.perform(webContext,
       config,
       config.getHttpActionAdapter.asInstanceOf[HttpActionAdapter[IO[Response[IO]], Http4sWebContext]],
       this.defaultUrl.orNull,
       this.saveInSession,
       this.multiProfile,
       this.renewSession,
-      this.defaultClient.orNull)
+      this.defaultClient.orNull)).flatten
   }
 }

--- a/src/main/scala/org/pac4j/http4s/LogoutService.scala
+++ b/src/main/scala/org/pac4j/http4s/LogoutService.scala
@@ -1,6 +1,7 @@
 package org.pac4j.http4s
 
-import cats.effect.IO
+import cats.implicits._
+import cats.effect._
 import org.http4s.{Response, _}
 import org.pac4j.core.config.Config
 import org.pac4j.core.engine.DefaultLogoutLogic
@@ -12,22 +13,24 @@ import org.pac4j.core.http.adapter.HttpActionAdapter
   * @author Iain Cardnell
   */
 class LogoutService(config: Config,
+                    blocker: Blocker,
                     defaultUrl: Option[String] = None,
                     logoutUrlPattern: Option[String] = None,
                     localLogout: Boolean = true,
                     destroySession: Boolean = false,
-                    centralLogout: Boolean = false) {
+                    centralLogout: Boolean = false)
+                   (implicit cs: ContextShift[IO], sync: Sync[IO]) {
 
   def logout(request: Request[IO]): IO[Response[IO]] = {
     val logoutLogic = new DefaultLogoutLogic[IO[Response[IO]], Http4sWebContext]()
     val webContext = Http4sWebContext(request, config)
-    logoutLogic.perform(webContext,
+    blocker.delay(logoutLogic.perform(webContext,
       config,
       config.getHttpActionAdapter.asInstanceOf[HttpActionAdapter[IO[Response[IO]], Http4sWebContext]],
       this.defaultUrl.orNull,
       this.logoutUrlPattern.orNull,
       this.localLogout,
       this.destroySession,
-      this.centralLogout)
+      this.centralLogout)).flatten
   }
 }

--- a/src/main/scala/org/pac4j/http4s/LogoutService.scala
+++ b/src/main/scala/org/pac4j/http4s/LogoutService.scala
@@ -19,12 +19,12 @@ class LogoutService(config: Config,
                     localLogout: Boolean = true,
                     destroySession: Boolean = false,
                     centralLogout: Boolean = false)
-                   (implicit cs: ContextShift[IO], sync: Sync[IO]) {
+                   (implicit cs: ContextShift[IO]) {
 
   def logout(request: Request[IO]): IO[Response[IO]] = {
     val logoutLogic = new DefaultLogoutLogic[IO[Response[IO]], Http4sWebContext]()
     val webContext = Http4sWebContext(request, config)
-    blocker.delay(logoutLogic.perform(webContext,
+    blocker.delay[IO, IO[Response[IO]]](logoutLogic.perform(webContext,
       config,
       config.getHttpActionAdapter.asInstanceOf[HttpActionAdapter[IO[Response[IO]], Http4sWebContext]],
       this.defaultUrl.orNull,

--- a/src/main/scala/org/pac4j/http4s/SecurityFilterMiddleware.scala
+++ b/src/main/scala/org/pac4j/http4s/SecurityFilterMiddleware.scala
@@ -2,7 +2,8 @@ package org.pac4j.http4s
 
 import java.util
 
-import cats.effect.IO
+import cats.implicits._
+import cats.effect._
 import org.http4s.server.{HttpMiddleware, Middleware}
 import org.http4s.{HttpRoutes, Response}
 import org.http4s.implicits._
@@ -43,6 +44,7 @@ object SecurityFilterMiddleware {
 
   def securityFilter(
       config: Config,
+      blocker: Blocker,
       clients: Option[String] = None,
       authorizers: Option[String] = None,
       matchers: Option[String] = None,
@@ -51,13 +53,13 @@ object SecurityFilterMiddleware {
         IO[Response[IO]],
         Http4sWebContext
       ] = new DefaultSecurityGrantedAccessAdapter(_)
-  ): HttpMiddleware[IO] =
+    )(implicit cs: ContextShift[IO], sync: Sync[IO]): HttpMiddleware[IO] =
     Middleware { (request, service) =>
       val securityLogic =
         new DefaultSecurityLogic[IO[Response[IO]], Http4sWebContext]
       val context = Http4sWebContext(request, config)
       OptionT.liftF(
-        securityLogic.perform(
+        blocker.delay(securityLogic.perform(
           context,
           config,
           securityGrantedAccessAdapter(service),
@@ -68,6 +70,6 @@ object SecurityFilterMiddleware {
           matchers.orNull,
           multiProfile
         )
-      )
+      ).flatten)
     }
 }

--- a/src/main/scala/org/pac4j/http4s/SecurityFilterMiddleware.scala
+++ b/src/main/scala/org/pac4j/http4s/SecurityFilterMiddleware.scala
@@ -53,13 +53,13 @@ object SecurityFilterMiddleware {
         IO[Response[IO]],
         Http4sWebContext
       ] = new DefaultSecurityGrantedAccessAdapter(_)
-    )(implicit cs: ContextShift[IO], sync: Sync[IO]): HttpMiddleware[IO] =
+    )(implicit cs: ContextShift[IO]): HttpMiddleware[IO] =
     Middleware { (request, service) =>
       val securityLogic =
         new DefaultSecurityLogic[IO[Response[IO]], Http4sWebContext]
       val context = Http4sWebContext(request, config)
       OptionT.liftF(
-        blocker.delay(securityLogic.perform(
+        blocker.delay[IO, IO[Response[IO]]](securityLogic.perform(
           context,
           config,
           securityGrantedAccessAdapter(service),

